# Answering Business Questions

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

Schema Diagram
Click here![links](http://www.basedonstatistics.com/wp-content/uploads/2018/06/chnk-768x761.png)

**Tasks:**
- First we will write a query that returns the table with thier informations
- explore the tables

In [2]:
%%sql
SELECT *
FROM sqlite_master
WHERE type IN ('table','view');

 * sqlite:///chinook.db
Done.


[('table', 'album', 'album', 2, 'CREATE TABLE [album]\n(\n    [album_id] INTEGER PRIMARY KEY NOT NULL,\n    [title] NVARCHAR(160)  NOT NULL,\n    [artist_id] INTEGER  NOT NULL,\n    FOREIGN KEY ([artist_id]) REFERENCES [artist] ([artist_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'artist', 'artist', 3, 'CREATE TABLE [artist]\n(\n    [artist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'customer', 'customer', 4, 'CREATE TABLE [customer]\n(\n    [customer_id] INTEGER PRIMARY KEY NOT NULL,\n    [first_name] NVARCHAR(40)  NOT NULL,\n    [last_name] NVARCHAR(20)   ... (276 characters truncated) ...  [support_rep_id] INTEGER,\n    FOREIGN KEY ([support_rep_id]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'employee', 'employee', 5, 'CREATE TABLE [employee]\n(\n    [employee_id] INTEGER PRIMARY KEY NOT NULL,\n    [last_name] NVARCHAR(20)  NOT NULL,\n    [first_name] NVARCHAR(20)   ... (310 characters truncated) ... 4),\n    [email] NVARCHAR(60),\n    FOREIGN KEY ([reports_to]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'genre', 'genre', 6, 'CREATE TABLE [genre]\n(\n    [genre_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'invoice', 'invoice', 7, 'CREATE TABLE [invoice]\n(\n    [invoice_id] INTEGER PRIMARY KEY NOT NULL,\n    [customer_id] INTEGER  NOT NULL,\n    [invoice_date] DATETIME  NOT NUL ... (195 characters truncated) ... tal] NUMERIC(10,2)  NOT NULL,\n    FOREIGN KEY ([customer_id]) REFERENCES [customer] ([customer_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'invoice_line', 'invoice_line', 8, 'CREATE TABLE [invoice_line]\n(\n    [invoice_line_id] INTEGER PRIMARY KEY NOT NULL,\n    [invoice_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT ... (161 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'media_type', 'media_type', 9, 'CREATE TABLE [media_type]\n(\n    [media_type_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist', 'playlist', 10, 'CREATE TABLE [playlist]\n(\n    [playlist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist_track', 'playlist_track', 11, 'CREATE TABLE [playlist_track]\n(\n    [playlist_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT NULL,\n    CONSTRAINT [pk_playlist_track] PRIMARY ... (114 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'track', 'track', 13, 'CREATE TABLE [track]\n(\n    [track_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(200)  NOT NULL,\n    [album_id] INTEGER,\n    [media_type_ ... (377 characters truncated) ... ON ON UPDATE NO ACTION,\n    FOREIGN KEY ([media_type_id]) REFERENCES [media_type] ([media_type_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('view', 'customer_usa', 'customer_usa', 0, 'CREATE VIEW customer_usa AS \n    SELECT * FROM customer\n     WHERE country = "USA"'),
 ('view', 'customer_get_90_dollars', 'customer_get_90_dollars', 0, 'CREATE VIEW customer_get_90_dollars AS\nSELECT *\nFROM invoice\nGROUP BY 1\nHAVING total < 90'),
 ('view', 'customer_gt_90_dollars', 'customer_gt_90_dollars', 0, 'CREATE VIEW customer_gt_90_dollars AS\nSELECT c.*\nFROM invoice i\nINNER JOIN customer c ON c.customer_id = i.customer_id\nGROUP BY 1\nHAVING SUM(total) > 90'),
 ('view', 'invoice_line_usa', 'invoice_line_usa', 0, 'CREATE VIEW invoice_line_usa AS\nSELECT  * \nFROM invoice i\nLEFT JOIN  invoice_line il ON il.invoice_id = i.invoice_id\nWHERE billing_country LIKE "%usa%"'),
 ('view', 'artist_album', 'artist_album', 0, 'CREATE VIEW artist_album AS\nSELECT *\nFROM artist ar\nLEFT JOIN alb

In [3]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM invoice;

 * sqlite:///chinook.db
Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 614)]

In [4]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM invoice_line;

 * sqlite:///chinook.db
Done.


[(1, 1, 1158, 0.99, 1, 4757)]

In [5]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM genre;

 * sqlite:///chinook.db
Done.


[(1, 'Rock', 25)]

In [6]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM track

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99, 3503)]

In [7]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM album;

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock We Salute You', 1, 347)]

In [8]:
%%sql
SELECT  * , COUNT(*) "number_of_rows"
FROM artist;

 * sqlite:///chinook.db
Done.


[(1, 'AC/DC', 275)]

#### Obeservation :
- **genre** : 2 columns and 25 rows
- **track** : 9 columns and 3503 rows.
- **album** : 3 columns and 347 rows.
- **artist** : 2 columns and 275 rows. 
- **invoice_line** : 5 columns and 4757 rows.
- **invoice** : 9 columns and 614 rows.

**Tasks:**
- so we're interested in finding out which genres sell the best in the USA.
- find out which genres sell the most tracks in the USA
- make a recommendation for the three artists whose albums we should purchase for the store.

> to answe these question we will need to work with multipule tables .

In [9]:
%%sql
SELECT *
FROM invoice_line_usa;

 * sqlite:///chinook.db
Done.


[(1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 1, 1, 1158, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 2, 1, 1159, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 3, 1, 1160, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 4, 1, 1161, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 5, 1, 1162, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 6, 1, 1163, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 7, 1, 1164, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 8, 1, 1165, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 9, 1, 1166, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 10, 1, 1167, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 11, 1, 1168, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 12, 1, 1169, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 13, 1, 1170, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 14, 1, 1171, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 15, 1, 1172, 0.99, 1),
 (1, 18, '2017-01-03 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 15.84, 16, 1, 1173, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 29, 4, 3448, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 30, 4, 2560, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 31, 4, 3336, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 32, 4, 829, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 33, 4, 1872, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 34, 4, 748, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 35, 4, 1778, 0.99, 1),
 (4, 18, '2017-01-06 00:00:00', '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', 7.92, 36, 4, 2514, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 37, 5, 1986, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 38, 5, 1987, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 39, 5, 1988, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 40, 5, 1989, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 41, 5, 1990, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 42, 5, 1991, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 43, 5, 1992, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 44, 5, 1993, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 45, 5, 1994, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', 16.83, 46, 5, 1995, 0.99, 1),
 (5, 27, '2017-01-07 00:00:00', '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719'

In [10]:
%%sql
SELECT *
FROM artist_album
ORDER BY album_id;

 * sqlite:///chinook.db
Done.


[(25, 'Milton Nascimento & Bebeto', None, None, None),
 (26, 'Azymuth', None, None, None),
 (28, 'João Gilberto', None, None, None),
 (29, 'Bebel Gilberto', None, None, None),
 (30, 'Jorge Vercilo', None, None, None),
 (31, 'Baby Consuelo', None, None, None),
 (32, 'Ney Matogrosso', None, None, None),
 (33, 'Luiz Melodia', None, None, None),
 (34, 'Nando Reis', None, None, None),
 (35, 'Pedro Luís & A Parede', None, None, None),
 (38, 'Banda Black Rio', None, None, None),
 (39, 'Fernanda Porto', None, None, None),
 (40, 'Os Cariocas', None, None, None),
 (43, 'A Cor Do Som', None, None, None),
 (44, 'Kid Abelha', None, None, None),
 (45, 'Sandra De Sá', None, None, None),
 (47, 'Hermeto Pascoal', None, None, None),
 (48, 'Barão Vermelho', None, None, None),
 (49, 'Edson, DJ Marky & DJ Patife Featuring Fernanda Porto', None, None, None),
 (60, 'Santana Feat. Dave Matthews', None, None, None),
 (61, 'Santana Feat. Everlast', None, None, None),
 (62, 'Santana Feat. Rob Thomas', None, None, None),
 (63, 'Santana Feat. Lauryn Hill & Cee-Lo', None, None, None),
 (64, 'Santana Feat. The Project G&B', None, None, None),
 (65, 'Santana Feat. Maná', None, None, None),
 (66, 'Santana Feat. Eagle-Eye Cherry', None, None, None),
 (67, 'Santana Feat. Eric Clapton', None, None, None),
 (71, 'Vinícius De Moraes & Baden Powell', None, None, None),
 (73, 'Vinícius E Qurteto Em Cy', None, None, None),
 (74, 'Vinícius E Odette Lara', None, None, None),
 (75, 'Vinicius, Toquinho & Quarteto Em Cy', None, None, None),
 (107, 'Motörhead & Girlschool', None, None, None),
 (119, 'Peter Tosh', None, None, None),
 (123, 'R.E.M. Feat. KRS-One', None, None, None),
 (129, 'Simply Red', None, None, None),
 (154, 'Whitesnake', None, None, None),
 (160, 'Christina Aguilera featuring BigElf', None, None, None),
 (161, "Aerosmith & Sierra Leone's Refugee Allstars", None, None, None),
 (162, 'Los Lonely Boys', None, None, None),
 (163, 'Corinne Bailey Rae', None, None, None),
 (164, 'Dhani Harrison & Jakob Dylan', None, None, None),
 (165, 'Jackson Browne', None, None, None),
 (166, 'Avril Lavigne', None, None, None),
 (167, 'Big & Rich', None, None, None),
 (168, "Youssou N'Dour", None, None, None),
 (169, 'Black Eyed Peas', None, None, None),
 (170, 'Jack Johnson', None, None, None),
 (171, 'Ben Harper', None, None, None),
 (172, 'Snow Patrol', None, None, None),
 (173, 'Matisyahu', None, None, None),
 (174, 'The Postal Service', None, None, None),
 (175, 'Jaguares', None, None, None),
 (176, 'The Flaming Lips', None, None, None),
 (177, "Jack's Mannequin & Mick Fleetwood", None, None, None),
 (178, 'Regina Spektor', None, None, None),
 (181, 'Xis', None, None, None),
 (182, 'Nega Gizza', None, None, None),
 (183, 'Gustavo & Andres Veiga & Salazar', None, None, None),
 (184, 'Rodox', None, None, None),
 (185, 'Charlie Brown Jr.', None, None, None),
 (186, 'Pedro Luís E A Parede', None, None, None),
 (187, 'Los Hermanos', None, None, None),
 (188, 'Mundo Livre S/A', None, None, None),
 (189, 'Otto', None, None, None),
 (190, 'Instituto', None, None, None),
 (191, 'Nação Zumbi', None, None, None),
 (192, 'DJ Dolores & Orchestra Santa Massa', None, None, None),
 (193, 'Seu Jorge', None, None, None),
 (194, 'Sabotage E Instituto', None, None, None),
 (195, 'Stereo Maracana', None, None, None),
 (239, 'Academy of St. Martin in the Fields, Sir Neville Marriner & William Bennett', None, None, None),
 (1, 'AC/DC', 1, 'For Those About To Rock We Salute You', 1),
 (2, 'Accept', 2, 'Balls to the Wall', 2),
 (2, 'Accept', 3, 'Restless and Wild', 2),
 (1, 'AC/DC', 4, 'Let There Be Rock', 1),
 (3, 'Aerosmith', 5, 'Big Ones', 3),
 (4, 'Alanis Morissette', 6, 'Jagged Little Pill', 4),
 (5, 'Alice In Chains', 7, 'Facelift', 5),
 (6, 'Antônio Carlos Jobim', 8, 'Warner 25 Anos', 6),
 (7, 'Apocalyptica', 9, 'Plays Metallica By Four Cellos', 7),
 (8, 'Audioslave', 10, 'Audioslave', 8),
 (8, 'Audioslave', 11, 'Out Of Exile', 8),
 (9, 'BackBeat', 12, 'BackBeat Soundtrack', 9),
 (10, 

**we have created two views.**

invoice_line_usa .
artist_album .

now it littel bit easier to work with three tables .
## Analyzing 

In [11]:
%%sql
WITH invoice_info AS
(
SELECT *
    FROM invoice_line_usa ilu
    LEFT JOIN track t ON t.track_id = ilu.track_id
)
SELECT gr.name "Genre", COUNT(if.name) "Number_of_Tracks",ROUND(COUNT(if.name),2)/(SELECT
                                                                                  ROUND(COUNT(*),2)
                                                                                   FROM invoice_line_usa
                                                                                  ) AS "Number_tracks_per"
FROM genre gr
LEFT JOIN invoice_info if ON if.genre_id = gr.genre_id 
GROUP BY gr.name
ORDER BY 2 DESC;

 * sqlite:///chinook.db
Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Pop', 22, 0.02093244529019981),
 ('Latin', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425),
 ('Easy Listening', 13, 0.012369172216936251),
 ('Reggae', 6, 0.005708848715509039),
 ('Electronica/Dance', 5, 0.004757373929590866),
 ('Classical', 4, 0.003805899143672693),
 ('Heavy Metal', 3, 0.0028544243577545195),
 ('Soundtrack', 2, 0.0019029495718363464),
 ('TV Shows', 1, 0.0009514747859181732),
 ('World', 0, 0.0),
 ('Science Fiction', 0, 0.0),
 ('Sci Fi & Fantasy', 0, 0.0),
 ('Rock And Roll', 0, 0.0),
 ('Opera', 0, 0.0),
 ('Drama', 0, 0.0),
 ('Comedy', 0, 0.0),
 ('Bossa Nova', 0, 0.0)]

OBSERVATION :
- based on sales of tracks from their genres sold in the USA i can recommend 
Alternative & Punk, blues and Pop wich all of them has high percentages   
- Rock have the highest sales in USA and has 0.53% of sales.


**Write a query that finds the total dollar amount of sales assigned to each sales support agent within the company.**

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. You have been asked to analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

You might like to consider whether any extra columns from the employee table explain any variance you see, or whether the variance might instead be indicative of employee performance.

**Instructions**

- Write a query that finds the total dollar amount of sales assigned to each sales support agent within the company. Add any extra attributes for that employee that you find are relevant to the analysis.
- Write a short statement describing your results, and providing a possible interpretation.

In [12]:
%%sql
SELECT *
FROM customer c
LEFT JOIN ( SELECT *
           FROM employee
           WHERE title LIKE "%Sales Support Agent%" 
           
          )e ON e.employee_id = c.support_rep_id


 * sqlite:///chinook.db
Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3, 3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5, 5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com'),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3, 3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4, 4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4, 4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5, 5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com'),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5, 5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com'),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4, 4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4, 4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4, 4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (11, 'Alexandre', 'Rocha', 'Banco do Brasil S.A.', 'Av. Paulista, 2022', 'São Paulo', 'SP', 'Brazil', '01310-200', '+55 (11) 3055-3278', '+55 (11) 3055-8131', 'alero@uol.com.

In [13]:
%%sql
WITH customer_info AS
    (
    SELECT *
        FROM customer c
        LEFT JOIN invoice i ON i.customer_id = c.customer_id
    ),
    employee_info AS
    (
    SELECT *
    FROM employee e
    WHERE title LIKE "%Sales Support Agent%" 
        )
SELECT 
        ei.first_name || " " || ei.last_name "Employee_name",
        SUM(total) "Total"

FROM customer_info ci
LEFT JOIN employee_info ei ON ei.employee_id = ci.support_rep_id
GROUP BY 1
ORDER BY 2 DESC;

 * sqlite:///chinook.db
Done.


[('Jane Peacock', 1731.510000000004),
 ('Margaret Park', 1584.0000000000032),
 ('Steve Johnson', 1393.9200000000028)]

Observation:
- Jane Peacock has the highest sales  1731.51 $.

- Margaret Park is the Second highest 1584 $.

- Steve Johnson	has 1393.92 $.

**Write a query that collates data on purchases from different countries**

Your next task is to analyze the sales data for customers from each different country. You have been given guidance to use the country value from the customers table, and ignore the country from the billing address in the invoice table.

In particular, you have been directed to calculate data, for each country, on the:

- total number of customers
- total value of sales
- average value of sales per customer
- average order value

In [14]:
%%sql
WITH customer_info AS 
    (
    SELECT  customer_id,  first_name || " " || last_name AS "customer_name",country
        FROM customer
    )

SELECT i.customer_id, 
    customer_name,country, 
    ROUND(SUM(total),2) "total_sales_per_country",
    COUNT(DISTINCT customer_name) "Number_of_customer_per_country"
FROM invoice i
INNER JOIN customer_info ci ON ci.customer_id = i.customer_id
GROUP BY 3
ORDER BY 5 DESC

 * sqlite:///chinook.db
Done.


[(18, 'Michelle Brooks', 'USA', 1040.49, 13),
 (30, 'Edward Francis', 'Canada', 535.59, 8),
 (40, 'Dominique Lefebvre', 'France', 389.07, 5),
 (1, 'Luís Gonçalves', 'Brazil', 427.68, 5),
 (38, 'Niklas Schröder', 'Germany', 334.62, 4),
 (53, 'Phil Hughes', 'United Kingdom', 245.52, 3),
 (35, 'Madalena Sampaio', 'Portugal', 185.13, 2),
 (59, 'Puja Srivastava', 'India', 183.15, 2),
 (5, 'František Wichterlová', 'Czech Republic', 273.24, 2),
 (51, 'Joakim Johansson', 'Sweden', 75.24, 1),
 (50, 'Enrique Muñoz', 'Spain', 98.01, 1),
 (49, 'Stanisław Wójcik', 'Poland', 76.23, 1),
 (4, 'Bjørn Hansen', 'Norway', 72.27, 1),
 (48, 'Johannes Van der Berg', 'Netherlands', 65.34, 1),
 (47, 'Lucas Mancini', 'Italy', 50.49, 1),
 (46, "Hugh O'Reilly", 'Ireland', 114.84, 1),
 (45, 'Ladislav Kovács', 'Hungary', 78.21, 1),
 (44, 'Terhi Hämäläinen', 'Finland', 79.2, 1),
 (9, 'Kara Nielsen', 'Denmark', 37.62, 1),
 (57, 'Luis Rojas', 'Chile', 97.02, 1),
 (8, 'Daan Peeters', 'Belgium', 60.39, 1),
 (7, 'Astrid Gruber', 'Austria', 69.3, 1),
 (55, 'Mark Taylor', 'Australia', 81.18, 1),
 (56, 'Diego Gutiérrez', 'Argentina', 39.6, 1)]

Number of unique customer per Country:
- the USA have the highest unique customer .

total sales per country 
- the USA has the highest sales worth about 1040.49 $. 

In [15]:
%%sql
WITH customer_info AS 
    (
    SELECT  customer_id,  first_name || " " || last_name AS "customer_name",country
        FROM customer
    )
    
SELECT i.customer_id, customer_name,country, ROUND(SUM(total),2) "total_sales_per_country",AVG(total) "avrage_sales_per_customer"
FROM invoice i
INNER JOIN customer_info ci ON ci.customer_id = i.customer_id
GROUP BY 2
ORDER BY 4 DESC;

 * sqlite:///chinook.db
Done.


[(5, 'František Wichterlová', 'Czech Republic', 144.54, 8.030000000000001),
 (6, 'Helena Holý', 'Czech Republic', 128.7, 10.725),
 (46, "Hugh O'Reilly", 'Ireland', 114.84, 8.833846153846151),
 (58, 'Manoj Pareek', 'India', 111.87, 8.605384615384615),
 (1, 'Luís Gonçalves', 'Brazil', 108.9, 8.376923076923076),
 (13, 'Fernanda Ramos', 'Brazil', 106.92, 7.127999999999999),
 (34, 'João Fernandes', 'Portugal', 102.96, 7.920000000000001),
 (42, 'Wyatt Girard', 'France', 99.99, 9.09),
 (3, 'François Tremblay', 'Canada', 99.99, 11.11),
 (53, 'Phil Hughes', 'United Kingdom', 98.01, 8.91),
 (17, 'Jack Smith', 'USA', 98.01, 8.1675),
 (50, 'Enrique Muñoz', 'Spain', 98.01, 8.91),
 (57, 'Luis Rojas', 'Chile', 97.02, 7.463076923076924),
 (20, 'Dan Miller', 'USA', 95.04, 7.919999999999999),
 (37, 'Fynn Zimmermann', 'Germany', 94.05, 9.405000000000001),
 (22, 'Heather Leacock', 'USA', 92.07, 7.6725),
 (21, 'Kathy Chase', 'USA', 91.08, 8.280000000000001),
 (30, 'Edward Francis', 'Canada', 91.08, 7.006153846153846),
 (26, 'Richard Cunningham', 'USA', 86.13, 7.177500000000002),
 (36, 'Hannah Schneider', 'Germany', 85.14, 7.74),
 (27, 'Patrick Gray', 'USA', 84.15, 9.35),
 (12, 'Roberto Almeida', 'Brazil', 82.17, 7.47),
 (35, 'Madalena Sampaio', 'Portugal', 82.17, 5.135625),
 (2, 'Leonie Köhler', 'Germany', 82.17, 7.47),
 (55, 'Mark Taylor', 'Australia', 81.18, 8.118),
 (44, 'Terhi Hämäläinen', 'Finland', 79.2, 7.2),
 (54, 'Steve Murray', 'United Kingdom', 79.2, 8.8),
 (18, 'Michelle Brooks', 'USA', 79.2, 9.9),
 (39, 'Camille Bernard', 'France', 79.2, 8.8),
 (45, 'Ladislav Kovács', 'Hungary', 78.21, 7.821),
 (25, 'Victor Stevens', 'USA', 76.23, 7.623),
 (49, 'Stanisław Wójcik', 'Poland', 76.23, 7.622999999999999),
 (51, 'Joakim Johansson', 'Sweden', 75.24, 7.523999999999999),
 (33, 'Ellie Sullivan', 'Canada', 75.24, 6.2700000000000005),
 (16, 'Frank Harris', 'USA', 74.25, 9.28125),
 (38, 'Niklas Schröder', 'Germany', 73.26, 8.139999999999999),
 (43, 'Isabelle Mercier', 'France', 73.26, 6.1049999999999995),
 (28, 'Julia Barnett', 'USA', 72.27, 7.226999999999999),
 (40, 'Dominique Lefebvre', 'France', 72.27, 8.03),
 (4, 'Bjørn Hansen', 'Norway', 72.27, 8.030000000000001),
 (59, 'Puja Srivastava', 'India', 71.28, 8.91),
 (24, 'Frank Ralston', 'USA', 71.28, 8.91),
 (32, 'Aaron Mitchell', 'Canada', 70.29, 8.786249999999999),
 (7, 'Astrid Gruber', 'Austria', 69.3, 7.699999999999999),
 (11, 'Alexandre Rocha', 'Brazil', 69.3, 6.93),
 (52, 'Emma Jones', 'United Kingdom', 68.31, 8.53875),
 (23, 'John Gordon', 'USA', 66.33, 6.633),
 (15, 'Jennifer Peterson', 'Canada', 66.33, 7.37),
 (48, 'Johannes Van der Berg', 'Netherlands', 65.34, 6.534000000000001),
 (41, 'Marc Dubois', 'France', 64.35, 7.1499999999999995),
 (31, 'Martha Silk', 'Canada', 62.37, 5.670000000000001),
 (10, 'Eduardo Martins', 'Brazil', 60.39, 5.0325),
 (8, 'Daan Peeters', 'Belgium', 60.39, 8.627142857142855),
 (19, 'Tim Goyer', 'USA', 54.45, 6.05),
 (47, 'Lucas Mancini', 'Italy', 50.49, 5.61),
 (29, 'Robert Brown', 'Canada', 40.59, 10.1475),
 (56, 'Diego Gutiérrez', 'Argentina', 39.6, 7.92),
 (9, 'Kara Nielsen', 'Denmark', 37.62, 3.761999999999999),
 (14, 'Mark Philips', 'Canada', 29.7, 2.9699999999999998)]

_________

the total sales per customer 
as we can see the highest sales per customer is František Wichterlová with total purchase 144.54 %
and an AVRAGE 8.030000000000001

**Write a query that categorizes each invoice as either an album purchase or not?**

The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:

purchase a whole album
purchase a collection of one or more individual tracks.
The store does not let customers purchase a whole album, and then add individual tracks to that same purchase (unless they do that by choosing each track manually). When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.



**Instructions:**

Write a query that categorizes each invoice as either an album purchase or not, and calculates the following summary statistics:

- Number of invoices
- Percentage of invoices
- Write one to two sentences explaining your findings, and making a prospective recommendation on whether the Chinook store should continue to buy full albums from record companies

In [16]:

%%sql
WITH invoice_first_track AS
    (
     SELECT
         il.invoice_id invoice_id,
         MIN(il.track_id) first_track_id
     FROM invoice_line il
     GROUP BY 1
    )

SELECT
    album_purchase,
    COUNT(invoice_id) number_of_invoices,
    CAST(count(invoice_id) AS FLOAT) / (
                                         SELECT COUNT(*) FROM invoice
                                      ) percent
FROM
    (
    SELECT
        ifs.*,
        CASE
            WHEN
                 (
                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 

                  EXCEPT 

                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id
                 ) IS NULL
             AND
                 (
                  SELECT il2.track_id FROM invoice_line il2
                  WHERE il2.invoice_id = ifs.invoice_id

                  EXCEPT 

                  SELECT t.track_id FROM track t
                  WHERE t.album_id = (
                                      SELECT t2.album_id FROM track t2 
                                      WHERE t2.track_id = ifs.first_track_id
                                     ) 
                 ) IS NULL
             THEN "yes"
             ELSE "no"
         END AS "album_purchase"
     FROM invoice_first_track ifs
    )
GROUP BY album_purchase;


 * sqlite:///chinook.db
Done.


[('no', 500, 0.8143322475570033), ('yes', 114, 0.18566775244299674)]

Album purchases account for 18.6% of purchases. Based on this data, I would recommend against purchasing only select tracks from albums from record companies, since there is potential to lose one fifth of revenue.